sample-controller: add example CRD controller

munnerz opened the pull request:
**What this PR does / why we need it**:

Adds a sample-controller example repository

fixes #52752

**Special notes for your reviewer**:

This is currently based on the sttts:sttts-codegen-scripts branch and should not be merged until that is (ref https://github.com/kubernetes/kubernetes/pull/52186)

**Release note**:

```
Add k8s.io/sample-controller repository as an example of how to write a CRD controller (github.com/kubernetes/sample-controller)
```

/cc @sttts @nikhita @colemickens

munnerz commented:
I've seen that I need to run `./hack/update-bazel.sh` to update the bazel BUILD files, but when I do it generates the following BUILD files:

```
wrote "pkg/generated/openapi/BUILD"
wrote "staging/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/apis/example/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/clientset/internalversion/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/clientset/internalversion/fake/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/clientset/internalversion/scheme/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/clientset/internalversion/typed/example/internalversion/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/clientset/internalversion/typed/example/internalversion/fake/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/clientset/versioned/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/clientset/versioned/fake/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/clientset/versioned/scheme/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/clientset/versioned/typed/example/v1/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/clientset/versioned/typed/example/v1/fake/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/informers/externalversions/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/informers/externalversions/example/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/informers/externalversions/example/v1/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/informers/externalversions/internalinterfaces/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/informers/internalversion/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/informers/internalversion/example/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/informers/internalversion/example/internalversion/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/informers/internalversion/internalinterfaces/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/listers/example/internalversion/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/apiserver/listers/example/v1/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/crd/clientset/versioned/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/crd/clientset/versioned/fake/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/crd/clientset/versioned/scheme/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/crd/clientset/versioned/typed/example/v1/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/crd/clientset/versioned/typed/example/v1/fake/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/crd/informers/externalversions/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/crd/informers/externalversions/example/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/crd/informers/externalversions/example/v1/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/crd/informers/externalversions/internalinterfaces/BUILD"
wrote "staging/src/k8s.io/code-generator/examples/crd/listers/example/v1/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/clientset/internalversion/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/clientset/internalversion/fake/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/clientset/internalversion/scheme/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/clientset/internalversion/typed/wardle/internalversion/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/clientset/internalversion/typed/wardle/internalversion/fake/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/clientset/versioned/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/clientset/versioned/fake/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/clientset/versioned/scheme/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/clientset/versioned/typed/wardle/v1alpha1/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/clientset/versioned/typed/wardle/v1alpha1/fake/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/informers/externalversions/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/informers/externalversions/internalinterfaces/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/informers/externalversions/wardle/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/informers/externalversions/wardle/v1alpha1/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/informers/internalversion/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/informers/internalversion/internalinterfaces/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/informers/internalversion/wardle/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/informers/internalversion/wardle/internalversion/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/listers/wardle/internalversion/BUILD"
wrote "staging/src/k8s.io/sample-apiserver/pkg/client/listers/wardle/v1alpha1/BUILD"
wrote "staging/src/k8s.io/sample-controller/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/apis/example/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/apis/example/install/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/apis/example/v1alpha1/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/client/clientset/versioned/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/client/clientset/versioned/fake/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/client/clientset/versioned/scheme/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/client/clientset/versioned/typed/example/v1alpha1/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/client/clientset/versioned/typed/example/v1alpha1/fake/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/client/informers/externalversions/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/client/informers/externalversions/example/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/client/informers/externalversions/example/v1alpha1/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/client/informers/externalversions/internalinterfaces/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/client/listers/example/v1alpha1/BUILD"
wrote "staging/src/k8s.io/sample-controller/pkg/signals/BUILD"
```

Should there be so many changes? I'm not sure if these were forgotten before, or if I shouldn't be generated BUILD files for generated packages.

nikhita commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,301 @@
+	}
+}
+
+func kubeConfig() (*rest.Config, error) {
```
can we use something like this?
https://github.com/kubernetes/apiextensions-apiserver/blob/2bfb07d318ed549813240d1165fcebad6250c666/examples/client-go/main.go#L41

I think that's how it's used in other examples as well.

munnerz replied:
👍  done

nikhita commented on staging/src/k8s.io/sample-controller/main.go:
```diff
@@ -0,0 +1,42 @@
+	"flag"
+
+	"github.com/golang/glog"
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
```
Maybe something like this? Or this here for some other reason?

```
// Uncomment the following line to load the gcp plugin (only required to authenticate against GKE clusters).
// _ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
```

munnerz replied:
👍  done

nikhita commented on staging/src/k8s.io/sample-controller/main.go:
```diff
@@ -0,0 +1,42 @@
+)
+
+func main() {
+	flag.Parse()
```
Not sure what is the best practice here but we can we create the client config (from `kubeconfig`) here? Since it's part of the setup and not a part of the "controller" as such.

munnerz replied:
👍  done

nikhita commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,301 @@
+
+	exampleclientset, err := versioned.NewForConfig(cfg)
+
+	if err != nil {
```
nit: no newline here above the if clause.

munnerz replied:
👍  done

nikhita commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,301 @@
+}
+
+func (c *Controller) updateNGINXStatus(nginx *example.NGINX, deployment *apps.Deployment) error {
+	update := nginx.DeepCopy()
```
nit: maybe `nginxCopy` instead of `update`?

munnerz replied:
👍  done

nikhita commented on staging/src/k8s.io/sample-controller/pkg/apis/example/register.go:
```diff
@@ -0,0 +1,21 @@
+package example
+
+const (
+	GroupName = "example.controller.code-generator.k8s.io"
```
how about only `example.controller.k8s.io`?

munnerz replied:
I chose this as the sampe-apiserver changes that @sttts has made use `example.apiserver.code-generator.k8s.io`. I'm happy to change this to whatever makes sense though :)

nikhita commented on staging/src/k8s.io/sample-controller/pkg/apis/example/v1alpha1/types.go:
```diff
@@ -0,0 +1,56 @@
+// +resource:path=nginx
+
+// NGINX is a specification for a NGINX resource
+type NGINX struct {
```
Don't we want internal types as well? :thinking:

nikhita replied:
And hence generating the `zz_generated.conversion.go` files as well.

munnerz replied:
So if users are creating a CRD controller, conversions and internal types should not be required. Internal types & the conversions between versions are only useful when creating a custom API server (and are only used by the custom apiserver, iiuc)

The internal API version & conversions would never actually be used in the controller, so I thought it best to omit them to show users what is/isn't required. Maybe we should include a note in `pkg/apis/example/v1alpha1/types.go` explaining why we don't use the other generators?

nikhita replied:
>  The internal API version & conversions would never actually be used in the controller, so I thought it best to omit them to show users what is/isn't required. Maybe we should include a note in pkg/apis/example/v1alpha1/types.go explaining why we don't use the other generators?

Makes sense. :+1:

nikhita commented on staging/src/k8s.io/sample-controller/artifacts/examples/crd.yaml:
```diff
@@ -0,0 +1,11 @@
+  group: example.controller.code-generator.k8s.io
+  version: v1alpha1
+  names:
+    kind: NGINX
```
nit: not sure about any best practices but can we have this as `nginx` instead of `NGINX`? That way the plural (`nginxs`) seems good. Else, we should probably make it `NGINXS`.

nikhita replied:
Or should it be `nginxes`? :)

munnerz replied:
I'm not sure on the best thing to do here - nginxs works, I suppose, but does sound odd. I've never seen nginxes before. I wonder what the 'official' plural of NGINX is 😆 

I did try changing the `kind` field in the CRD to `nginx`, but upon running the controller get:

```k8s.io/kubernetes/vendor/k8s.io/sample-controller/pkg/client/informers/externalversions/factory.go:61: Failed to list *v1alpha1.NGINX: no kind "nginxList" is registered for version "example.controller.code-generator.k8s.io/v1alpha1"```

I've not had much time to dig into this right now, perhaps it's down to me still using the struct name NGINX and NGINXList in types.go?

I'm half thinking we should rename our resource to FooBar or something, although I'm not too sure...

nikhita replied:
> perhaps it's down to me still using the struct name NGINX and NGINXList in types.go?

Yup, should be just this.

> I'm half thinking we should rename our resource to FooBar or something, although I'm not too sure...

Yeah, I suppose a rename would be nice. :)

nikhita reviewed (commented):
Thanks for this! :tada: 

Left a few questions.

nikhita commented:
It would be nice if we could add a `README` and some more comments (examples: https://github.com/kubernetes/apiextensions-apiserver/blob/2bfb07d318ed549813240d1165fcebad6250c666/examples/client-go/controller/controller.go#L113, https://github.com/kubernetes/apiextensions-apiserver/blob/2bfb07d318ed549813240d1165fcebad6250c666/examples/client-go/controller/controller.go#L86 ) too! :)

nikhita commented:
@munnerz Some of those `BUILD` file changes will probably be picked up by https://github.com/kubernetes/kubernetes/pull/52186 later.

munnerz commented:
@nikhita I've gone through and added comments throughout, as well as addressing your feedback! Thanks for reviewing 😄 

I still need to put together a README, as well as changing the CI to include a verify of the generated code on each push. Any pointers to the best place to add this?

nikhita commented:
> Any pointers to the best place to add this?

I am not sure about this too but @sttts should know.

munnerz commented:
@nikhita managed to work out the code generation stuff! Fingers crossed that passes CI okay now (once some ok-to-test's this).

Final piece is a README, and renaming NGINX->(nginx|FooBar)

nikhita commented on staging/src/k8s.io/sample-controller/pkg/client/clientset/versioned/clientset.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2017 The Kubernetes Authors.
+Copyright 2017 The Kubernetes sample-controller Authors.
```
This needs to be `Copyright 2017 The Kubernetes Authors.` everywhere and not `Copyright 2017 The Kubernetes <repo> Authors.`

nikhita replied:
Hmm but these are generated files...

munnerz replied:
Not sure why these have been marked as outdated by GitHub.... 🤔 

I copied this from the sample-apiserver repo: https://github.com/sttts/kubernetes/blob/sttts-codegen-scripts/staging/src/k8s.io/sample-apiserver/hack/custom-boilerplate.go.txt

munnerz commented:
Rebased on latest changes to @sttts branch. I've also renamed NGINX to Foo, and added a README. I think this is ready now.

@nikhita @sttts please take a look in your own time!

munnerz commented:
Also let me know if a squashed version is easier to review now. I've got a local version that's condensed to 4 commits 😄

sttts commented on staging/src/k8s.io/sample-controller/hack/update-codegen.sh:
```diff
@@ -0,0 +1,25 @@
+CODEGEN_PKG=${CODEGEN_PKG:-$(cd ${SCRIPT_ROOT}; ls -d -1 ./vendor/k8s.io/code-generator 2>/dev/null || echo ../code-generator)}
+source ${CODEGEN_PKG}/hack/lib/codegen.sh
+
+codegen::generate-groups "deepcopy,client,informer,lister" "$(dirname ${BASH_SOURCE})/../../.." k8s.io/sample-controller/pkg/client k8s.io/sample-controller/pkg/apis example:v1alpha1 --go-header-file ${SCRIPT_ROOT}/hack/custom-boilerplate.go.txt
```
the output-base `"$(dirname ${BASH_SOURCE})/../../.."` shouldn't be necessary anymore

sttts replied:
I commented `--go-header-file ${SCRIPT_ROOT}/hack/custom-boilerplate.go.txt` out in sample-apiserver because otherwise the global kube code-gen scripts override our generated files with the default boilerplate. Would like to have a way to exclude repos from the global scripts.

munnerz replied:
Updated to reflect both these comments.

The `update-codegen.sh` script now looks similar to the one in `sample-apiserver`

munnerz replied:
How come the global scripts overwrite it? `./hack/update-codegen.sh` (in the root of the repo) simply calls out to the `update-codegen.sh` script in each staging repo. Unless you mean a different global script?

sttts replied:
Makefile.generated_files.

sttts commented on staging/src/k8s.io/sample-controller/pkg/apis/example/install/install.go:
```diff
@@ -0,0 +1,43 @@
+)
+
+// Install registers the API group and adds types to a scheme
+func Install(groupFactoryRegistry announced.APIGroupFactoryRegistry, registry *registered.APIRegistrationManager, scheme *runtime.Scheme) {
```
do we need install.go at all?

sttts commented on hack/verify-codegen.sh:
```diff
@@ -30,6 +30,7 @@ kube::golang::setup_env
 CODEGEN_PKG=./vendor/k8s.io/code-generator vendor/k8s.io/code-generator/hack/verify-codegen.sh
 CODEGEN_PKG=./vendor/k8s.io/code-generator vendor/k8s.io/kube-aggregator/hack/verify-codegen.sh
 CODEGEN_PKG=./vendor/k8s.io/code-generator vendor/k8s.io/sample-apiserver/hack/verify-codegen.sh
+CODEGEN_PKG=./vendor/k8s.io/code-generator vendor/k8s.io/sample-controller/hack/verify-codegen.sh
```
In fact, I think we can get rid of CODEGEN_PKG here now. But that's for another follow-up PR.

sttts commented on staging/src/k8s.io/sample-controller/README.md:
```diff
@@ -0,0 +1,24 @@
+This repository implements a simple controller for watching Foo resources as
+defined with a CustomResourceDefinition (CRD).
+
+It makes use of the generators in [k8s.io/code-generate](https://github.com/kubernetes/code-generator)
```
code-generator

sttts commented on staging/src/k8s.io/sample-controller/README.md:
```diff
@@ -0,0 +1,24 @@
+defined with a CustomResourceDefinition (CRD).
+
+It makes use of the generators in [k8s.io/code-generate](https://github.com/kubernetes/code-generator)
+to generate a typed client, informers, listers and deep copy functions. You can
```
deep-copy

sttts commented on staging/src/k8s.io/sample-controller/artifacts/examples/crd.yaml:
```diff
@@ -0,0 +1,11 @@
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: foos.example.controller.code-generator.k8s.io
```
how about foos.sample-controller.k8s.io ?

munnerz replied:
This name is invalid, as it'd result in the name of the group being `sample-controller`, which is invalid in Go package names. When building with this set, I get:

```
Generating clientset for example:v1alpha1 at k8s.io/sample-controller/pkg/client/clientset
F0925 21:59:07.859468   10700 main.go:202] Error: Failed executing generator: some packages had errors:
errors in package "k8s.io/sample-controller/pkg/client/clientset/versioned":
unable to format file "../../k8s.io/sample-controller/pkg/client/clientset/versioned/clientset.go" (24:8: expected 'STRING', found '-' (and 10 more errors)).

errors in package "k8s.io/sample-controller/pkg/client/clientset/versioned/scheme":
unable to format file "../../k8s.io/sample-controller/pkg/client/clientset/versioned/scheme/register.go" (22:8: expected 'STRING', found '-' (and 4 more errors)).
```

(with a load more, similar errors)

munnerz replied:
Okay I've managed to actually get it to work by setting the groupName to not include the hyphen (`// +groupName=samplecontroller.k8s.io`) - this then creates random inconsistencies between using the `-` and not using it. What do you think of `foos.samplecontroller.k8s.io`?

sttts replied:
without the `-` is fine

sttts commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,378 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	example "k8s.io/sample-controller/pkg/apis/example/v1alpha1"
```
examplev1alpha1, that's the convention within kube.

sttts commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,378 @@
+	"k8s.io/client-go/util/workqueue"
+
+	example "k8s.io/sample-controller/pkg/apis/example/v1alpha1"
+	"k8s.io/sample-controller/pkg/client/clientset/versioned"
```
would just call this clientset as alias

sttts commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,378 @@
+
+	example "k8s.io/sample-controller/pkg/apis/example/v1alpha1"
+	"k8s.io/sample-controller/pkg/client/clientset/versioned"
+	intinformers "k8s.io/sample-controller/pkg/client/informers/externalversions"
```
maybe just informers as alias?

sttts commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,378 @@
+
+type Controller struct {
+	// clientset is a standard kubernetes clientset
+	clientset kubernetes.Interface
```
kubeclient ?

sttts replied:
or kubeclientset

sttts commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,378 @@
+	exampleclientset versioned.Interface
+
+	// informerFactory produces shared informers for kubernetes types
+	informerFactory informers.SharedInformerFactory
```
would also prefix this with kube

sttts replied:
the informer factories shouldn't be part of this struct, but instatiated and started outside. They are meant to be shared if you have multiple controllers per binary.

sttts commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,378 @@
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	glog.Info("Starting workers...")
```
nit: no ...

sttts commented on staging/src/k8s.io/sample-controller/main.go:
```diff
@@ -0,0 +1,67 @@
+	cfg, err := buildConfig(kubeconfig)
+
+	if err != nil {
+		glog.Fatalf("error building kubeconfig: %s", err.Error())
```
nit: Error

sttts commented on staging/src/k8s.io/sample-controller/main.go:
```diff
@@ -0,0 +1,67 @@
+	}
+
+	controller, err := NewController(cfg)
+
```
nit: no empty line

sttts commented on staging/src/k8s.io/sample-controller/main.go:
```diff
@@ -0,0 +1,67 @@
+	flag.Parse()
+
+	cfg, err := buildConfig(kubeconfig)
+
```
nit: no empty line

sttts commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,378 @@
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			c.worker()
```
the worker call should be inside `wait.Until`, otherwise we won't recover from panic. Compare https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md

sttts commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,378 @@
+// worker is a long-running function that will continually read messages off
+// workqueue and attempt to process them. It will run until the workqueue has
+// been shutdown by a call to workqueue.ShutDown().
+func (c *Controller) worker() {
```
compare https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md for a best-practice structure of this.

sttts commented:
@munnerz I noticed that we divert quite a lot from https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md, the official controller guide. Can you try to adapt your code to that? E.g. starting of informers should happen outside and the worker structure is different.

sdminonne commented:
/cc

sttts commented:
@munnerz thanks for the changes. Looks pretty good now.

gregory-m commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,365 @@
+	// number does not equal the current desired replicas on the Deployment, we
+	// should update the Deployment resource.
+	if foo.Spec.Replicas != nil && *foo.Spec.Replicas != *deployment.Spec.Replicas {
+		deployment, err = c.kubeclientset.AppsV1beta1().Deployments(foo.Namespace).Update(newDeployment(foo))
```
I think deployment owner should be checked before update.

sttts commented:
@munnerz my code-generator PR merged.

nikhita commented on staging/src/k8s.io/sample-controller/main.go:
```diff
@@ -0,0 +1,86 @@
+	// set up signals so we handle the first shutdown signal gracefully
+	stopCh := signals.SetupSignalHandler()
+
+	cfg, err := buildConfig(kubeconfig)
```
Would be nice to refractor as per https://github.com/kubernetes/kubernetes/pull/53308.

munnerz committed:
Add sample CustomResourceDefinition controller

sample-controller: add API types

sample-controller: regenerate files

sample-controller: add implementation

sample-controller: update bazel

sample-controller: update make-rules and cache_go_dirs

sample-controller: Set noStatus tag. Remove openapi-gen tag.

sample-controller: add deletion tombstone handling logic

sample-controller: update README, remove use of reflect in UpdateFuncs

sample-controller: add OWNERS file

sample-controller: add LICENSE

munnerz commented:
/ok-to-test

munnerz commented:
I've rebased this and addressed the above comments.

I've also added a few more log messages, and added an event recorder to record the event where an existing deployment exists but is not owned by the Foo being processed, as well as recording successful sync events.

PTAL @nikhita @sttts @gregory-m

munnerz commented:
I've also run the `hack/update-bazel.sh` script. I've noticed it also touched `pkg/generated/openapi/BUILD`, which `sample-apiserver` does not feature in.

Does anyone have any idea if this is okay/correct, and if not, how to fix it?

sttts commented on staging/src/k8s.io/sample-controller/pkg/apis/samplecontroller/v1alpha1/types.go:
```diff
@@ -0,0 +1,57 @@
+)
+
+// +genclient
+// +k8s:openapi-gen=true
```
can you remove the openapi-gen tag? Our openapi generation code picks up this package, but shouldn't.

sttts commented on pkg/generated/openapi/BUILD:
```diff
@@ -61,6 +61,7 @@ openapi_library(
         "k8s.io/metrics/pkg/apis/custom_metrics/v1beta1",
         "k8s.io/metrics/pkg/apis/metrics/v1alpha1",
         "k8s.io/metrics/pkg/apis/metrics/v1beta1",
+        "k8s.io/sample-controller/pkg/apis/samplecontroller/v1alpha1",
```
with the openapi-tag removed from types.go this line will go.

sttts commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,413 @@
+	// Or create a copy manually for better performance
+	fooCopy := foo.DeepCopy()
+	fooCopy.Status.AvailableReplicas = deployment.Status.AvailableReplicas
+	// Until #38113 is merged, we must use Update instead of UpdateStatus to
```
UpdateStatus is just created in the client, but has no server counterpart (until we have subresources for CRDs, proposal is being discussed).

To get rid of UpdateStatue here, please use the noStatus tag, compare https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_type.go#L535.

munnerz replied:
Done

sttts commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,413 @@
+func (c *Controller) handleObject(obj interface{}) {
+	var object metav1.Object
+	var ok bool
+	if object, ok = obj.(metav1.Object); !ok {
```
for the delete case we get a tombstone here, i.e. the cast will fail. This need special logic in the informer event handler.

munnerz replied:
I've updated this accordingly. Please check over the changes here to make sure they look okay: https://github.com/kubernetes/kubernetes/pull/52753/commits/9151172943a50f28b16735e9e29890012cdc7017

munnerz commented:
@sttts I've addressed your comments and added two extra commits on top to make review easier. I can re-squash these once you're happy with the changes.

stevekuznetsov commented:
@kargakis @spxtr

sttts commented:
Looks good now. Leaving it for final review from @kargakis @spxtr. They wrote many controllers before.

kargakis commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,421 @@
+	SuccessSynced     = "Synced"
+	ErrResourceExists = "ErrResourceExists"
+
+	MessageResourceExists = "Resource '%s' already exists and is not managed by Foo"
```
nit: you can use %q instead of '%s'

kargakis commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,421 @@
+	fooInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.enqueueFoo,
+		UpdateFunc: func(old, new interface{}) {
+			if !reflect.DeepEqual(old, new) {
```
Usually, events on the main resource are handled unconditionally.

kargakis commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,421 @@
+	deploymentInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.handleObject,
+		UpdateFunc: func(old, new interface{}) {
+			if !reflect.DeepEqual(old, new) {
```
Events on secondary resources are handled based on a resource version comparison which is less expensive than a deep equality check, eg. https://github.com/kubernetes/kubernetes/blob/0564d5298504cb57ec4d35cf4878e383f9504ed6/pkg/controller/deployment/deployment_controller.go#L257

kargakis commented on staging/src/k8s.io/sample-controller/README.md:
```diff
@@ -0,0 +1,24 @@
+
+It makes use of the generators in [k8s.io/code-generator](https://github.com/kubernetes/code-generator)
+to generate a typed client, informers, listers and deep-copy functions. You can
+do this yourself using the `./hack/update-codegen.sh` script.
```
Can you add in the README what files do I need to have before running the script and what files will be generated by the script?

munnerz commented:
@kargakis thanks for the review. I've updated in line with all of your comments if you'd like to recheck. I can squash this down again afterwards.

munnerz commented:
@sttts once this is merged, what's the process for getting a new repository set up under the `kubernetes` org?

Also are there any changes to OWNERS files etc. to be made as part of this PR? I'm not sure how the parts of the codebase in `staging/` are supposed to be owned.

sttts commented:
> Also are there any changes to OWNERS

Yes, please create one. Can you put the two of us as owners? Plus probably everybody reviewed here in the PR as reviewer.

>  once this is merged, what's the process for getting a new repository set up under the kubernetes org?

I fear it's not well defined. There are some people who can create repos permission-wise (eparis, thockin, lavalamp and probably some more). It looks like an announcement mail to kube-devel is expected. If nobody speaks up, one of those people can create the repo. Then we need a change to https://github.com/kubernetes/k8s.io/blob/31a3045f626f09670fa997f1b2355c22182919e0/k8s.io/configmap-www-golang.yaml to make the repo go-gettable. The publisher bot is broken and maybe being fixed soon https://github.com/kubernetes/test-infra/pull/4134. It has a hard-coded repo list in Go code.

munnerz commented:
Great - I'll create an issue for it once this PR has been merged then too so we can track what needs to be done.

sttts commented:
Also don't forget a LICENSE and README file, similar to the other repos.

munnerz commented:
@sttts the sample-controller, code-generator and apiserver repos themselves don't contain a LICENSE nor a README. I think they only exist in the target repositories and not in the main `kubernetes/kubernetes` repo.

Should I add them here for now, and when we get a separate repo setup we can remove them from the `kuberenetes/kubernetes` repo?

nikhita commented on staging/src/k8s.io/sample-controller/OWNERS:
```diff
@@ -0,0 +1,7 @@
+reviewers:
+- gregory-m
+- kargakis
+- nikhita
```
Oops, I am not a kubernetes org member. :/

munnerz replied:
Do you need to be an org member to be added here?

Otherwise you can take a look at how to join the org here - https://github.com/kubernetes/community/blob/master/community-membership.md it's quite straightforward and I'm sure there are many that would sponsor you!

nikhita replied:
> Do you need to be an org member to be added here?

Yes.

Thanks, I'll consider applying! :)

nikhita replied:
@munnerz Now I am member. :+1:

sttts replied:
@munnerz ^^ now you can readd her

kargakis commented:
LGTM

kargakis commented:
Please squash and then I can tag

munnerz commented:
@kargakis do you want this squashed into a single commit? It's now 3 commits (vendor, generated files and implementation)

kargakis commented:
/lgtm

munnerz commented:
@kargakis  we just had a human race condition. I squashed the PR into a single commit and pushed about 1 second before you `/lgtm`. Letting you know in case that's an issue.

kargakis commented:
@munnerz that's ok, this still lgtm

munnerz commented:
@kargakis tests are currently failing due to this change https://github.com/kubernetes/kubernetes/pull/52753/files#diff-cb594faca05793c6495fbef52caf9b80R99

(failing because there are no tests for sample-controller yet, so 0 args are passed to `dirname`)

I've updated the PR to not include this change so it can be added alongside tests in future.

munnerz commented:
Sorry, previous link is broken. Should be: https://github.com/munnerz/kubernetes/commit/608343c208d8cad58bca9940822d690b05c6d397#diff-cb594faca05793c6495fbef52caf9b80R101

sttts commented on staging/src/k8s.io/sample-controller/OWNERS:
```diff
@@ -0,0 +1,6 @@
+- munnerz
+reviewers:
+- gregory-m
+- kargakis
```
I believe the approvers must be in the reviewer section as well.

sttts commented:
/lgtm

munnerz commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,431 @@
+		runtime.HandleError(err)
+		return
+	}
+	c.workqueue.Add(key)
```
Should I be using `AddRateLimited` here instead of `Add`? Given we set up a rate limited workqueue.

munnerz replied:
@kargakis @sttts

kargakis replied:
Ratelimiting should be done in case we requeue because of an error. Which reminds me that you have not added requeueing in case of errors returned from `syncHandler`.

kargakis commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,431 @@
+		// Run the syncHandler, passing it the namespace/name string of the
+		// Foo resource to be synced.
+		if err := c.syncHandler(key); err != nil {
+			return fmt.Errorf("error syncing '%s': %s", key, err.Error())
```
You need to add the key back to the queue here with ratelimiting. At least, that's what most if not all of the core controllers are doing since most of the errors are transient.

munnerz replied:
I didn't think this was required with a rate limit queue (explicitly putting the item back on the queue).

By not calling `Forget` on the key here due to the error, I thought the item was automatically requeued? (we instead only call `Done` a few lines above with `defer c.workqueue.Done(obj)`).

I'll update the Add call to use `AddRateLimited`. Please do correct me if I'm wrong above as that's how I've designed most of my controllers!

kargakis replied:
cc @deads2k

grdl replied:
@munnerz  I think either your assumption is wrong or there's something wrong with how rate limiting queue is implemented. If you don't add the key back it won't be requeued. Here's a simple test:
```
queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "TEST")
queue.Add("stuff")

got, _ := queue.Get()
fmt.Printf("%s\n", got)
queue.Done(got)

got2, _ := queue.Get()
fmt.Printf("%s\n", got2)
```
The sesond `Get()` will never return even though we didn't call `Forget()` and we expect the object to still be in the queue.

munnerz commented:
I'm also currently struggling to make tests pass due to `verify-staging-godeps` in CI removing a dependency from the `Godeps.json` generated locally. I've tried running a `git clean -fdx` to clean up any local generated files, to no avail.

The build log is here: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/52753/pull-kubernetes-verify/55748/

If you've any ideas I'd appreciate it! I'll continue to mess around a bit with it here 🤔

sttts commented:
> If you've any ideas I'd appreciate it! I'll continue to mess around a bit with it here 🤔

Have run `hack/update-staging-godeps.sh` here locally against your branch, without any changes. This is odd.

sttts commented:
@munnerz can you rebase? I think swagger 1.2 support was removed some days ago. You PR is merged with current master in CI. So you see the remove dep in CI, but not locally.

sttts commented on staging/src/k8s.io/sample-controller/hack/verify-codegen.sh:
```diff
@@ -0,0 +1,49 @@
+set -o pipefail
+
+SCRIPT_ROOT=$(dirname "${BASH_SOURCE}")/..
+SCRIPT_BASE=${SCRIPT_ROOT}/../..
```
SCRIPT_BASE is not used

munnerz commented:
@sttts Done and done

sttts commented on staging/src/k8s.io/sample-controller/pkg/apis/samplecontroller/v1alpha1/types.go:
```diff
@@ -0,0 +1,57 @@
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +resource:path=foo
```
Just seeing this tag the first time. For what is it? Looking through kube, it's not used, only in the k8s.io/metrics package.

munnerz replied:
I'm actually not 100% sure - I think this is something I've picked up from another package/example, could well be the `k8s.io/metrics` one!

I'll remove it now.

sttts commented on staging/src/k8s.io/sample-controller/main.go:
```diff
@@ -0,0 +1,80 @@
+
+	exampleClient, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		glog.Fatalf("Error building example clientset: %s", err.Error())
```
more idiomatic: `glog.Errorf("....: %v", err)`

mikedanese commented:
It seems like this would be more valuable if the sample-controller controlled apis hosted by the smaple apiserver.

sttts commented:
> It seems like this would be more valuable

This is a second use-case and yes, we should probably have an example, maybe just inside sample-apiserver. This is explicitly directed to CRD users.

mikedanese commented:
Ok, either way sounds fine.

chancez commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,431 @@
+	deploymentInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.handleObject,
+		UpdateFunc: func(old, new interface{}) {
+			newDepl := old.(*appsv1beta2.Deployment)
```
Should be using `new` not `old`: `newDepl := new.(*appsv1beta2.Deployment)`

gregory-m commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,431 @@
+	kubeclientset kubernetes.Interface,
+	exampleclientset clientset.Interface,
+	kubeInformerFactory kubeinformers.SharedInformerFactory,
+	exampleInformerFactory informers.SharedInformerFactory) (*Controller, error) {
```
You don't return error in any case inside this function.
Maybe error should be removed from function signature.

gregory-m commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,431 @@
+	recorder record.EventRecorder
+}
+
+// NewController returns a new controller for the given kubeconfig
```
You don't pass kubeconfig to this function.

CaoShuFeng commented on staging/src/k8s.io/sample-controller/controller.go:
```diff
@@ -0,0 +1,431 @@
+type Controller struct {
+	// kubeclientset is a standard kubernetes clientset
+	kubeclientset kubernetes.Interface
+	// exampleclientset is a clientset for our own API group
```
This docstring is out of date: 
s/exampleclientset/sampleclientset

smarterclayton commented:
/approve

sttts commented:
/retest
/lgtm

munnerz commented:
/retest

nikhita commented:
After this repo is published as `k8s.io/sample-controller`, should we remove the example in `apiextensions-apiserver` and point to this? The example there currently uses a dynamic client instead of a typed client, which is not a best practice.

sttts commented:
> which is not a best practice.

At least we should remove the controller part. We could keep a simple (typed) client example.

jennybuckley commented on staging/src/k8s.io/sample-controller/README.md:
```diff
@@ -0,0 +1,34 @@
+`sample-controller` is synced from
+https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/sample-controller.
+Code changes are made in that location, merged into k8s.io/kubernetes and
+later synced here.
```
@munnerz I'm confused about what this line means... changes are made here then synced to k8s.io/kubernetes then synced back to here again?
